Spark history read cannot deserialize metrics after FSCK Repair Run

# Environment
Local, S3

**Delta-rs version**: 0.24.0

**Binding**: Python

**Environment**:
- **Cloud provider**: Happens locally and on S3
- **OS**: NA
- **Other**: NA

***
# Bug

**What happened**:
Following an FSCK repair run from this library, a follow up lookup on the table history via Spark fails with the following error:
```
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `java.lang.String` from Array value (token `JsonToken.START_ARRAY`)
```

I fixed a similar bug with optimize runs early last year: https://github.com/delta-io/delta-rs/pull/2317

So I believe the fix here is similar (looking into it, but no luck so far). 

I believe the specific issue is due to the repair run outputting `files_removed` as an array and not a string: 
```
"files_removed":["part-00001-aa65f683-d845-408a-aa3d-2a3bd04e0f8a-c000.snappy.parquet"]
```

**What you expected to happen**: Following a repair run, delta-rs will write out the `files_removed` metric in a way Spark understands.

**How to reproduce it**:

Here is an MRE:

```
from deltalake import DeltaTable, write_deltalake
import pandas as pd
import time

# write some data into a delta table
df = pd.DataFrame({"id": [1, 2], "value": ["foo", "boo"]})
write_deltalake("./data/delta", df, mode="append")
# second write so there's a file to delete manually
write_deltalake("./data/delta", df, mode="append")

# Load data from the delta table
dt = DeltaTable("./data/delta")

print("Take a moment to delete one of the parquet files written out")
time.sleep(10)

dt.repair()

### SPARK SECTION ###
from pyspark.sql import SparkSession
from delta.tables import DeltaTable

# Initialize Spark session with Delta support
spark = SparkSession.builder \
    .appName("DeltaTableHistory") \
    .config("spark.jars.packages", "io.delta:delta-spark_2.12:3.2.1,org.apache.hadoop:hadoop-aws:3.2.1") \
    .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension") \
    .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog") \
    .getOrCreate()

delta_table_path = "./data/delta"
delta_table = DeltaTable.forPath(spark, delta_table_path)

# Get the history of the Delta table, fails here
history_df = delta_table.history()
```

**More details**:

From what I recall, I understand that delta-rs is technically following protocol spec. However from a potentially selfish perspective, this bug caused a breaking change in our Spark streaming jobs (which use table history under the hood) and I believe the best solution for now is to write this out in a way Spark understands.

--- a/crates/core/src/operations/filesystem_check.rs
+++ b/crates/core/src/operations/filesystem_check.rs
@@ -22,7 +22,7 @@ use futures::future::BoxFuture;
 use futures::StreamExt;
 pub use object_store::path::Path;
 use object_store::ObjectStore;
-use serde::Serialize;
+use serde::{de::Error as DeError, Deserialize, Deserializer, Serialize, Serializer};
 use url::{ParseError, Url};
 use uuid::Uuid;
 
@@ -56,6 +56,10 @@ pub struct FileSystemCheckMetrics {
     /// Was this a dry run
     pub dry_run: bool,
     /// Files that wrere removed successfully
+    #[serde(
+        serialize_with = "serialize_vec_string",
+        deserialize_with = "deserialize_vec_string"
+    )]
     pub files_removed: Vec<String>,
 }
 
@@ -64,6 +68,24 @@ struct FileSystemCheckPlan {
     log_store: LogStoreRef,
     /// Files that no longer exists in undlying ObjectStore but have active add actions
     pub files_to_remove: Vec<Add>,
+}
+
+// Custom serialization function that serializes metric details as a string
+fn serialize_vec_string<S>(value: &Vec<String>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let json_string = serde_json::to_string(value).map_err(serde::ser::Error::custom)?;
+    serializer.serialize_str(&json_string)
+}
+
+// Custom deserialization that parses a JSON string into MetricDetails
+fn deserialize_vec_string<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    serde_json::from_str(&s).map_err(DeError::custom)
 }
 
 fn is_absolute_path(path: &str) -> DeltaResult<bool> {

--- a/python/deltalake/table.py
+++ b/python/deltalake/table.py
@@ -69,6 +69,8 @@ MAX_SUPPORTED_READER_VERSION = 3
 NOT_SUPPORTED_READER_VERSION = 2
 SUPPORTED_READER_FEATURES = {"timestampNtz"}
 
+FSCK_METRICS_FILES_REMOVED_LABEL = "files_removed"
+
 FilterLiteralType = Tuple[str, str, Any]
 FilterConjunctionType = List[FilterLiteralType]
 FilterDNFType = List[FilterConjunctionType]
@@ -1430,7 +1432,11 @@ class DeltaTable:
             commit_properties,
             post_commithook_properties,
         )
-        return json.loads(metrics)
+        deserialized_metrics = json.loads(metrics)
+        deserialized_metrics[FSCK_METRICS_FILES_REMOVED_LABEL] = json.loads(
+            deserialized_metrics[FSCK_METRICS_FILES_REMOVED_LABEL]
+        )
+        return deserialized_metrics
 
     def transaction_versions(self) -> Dict[str, Transaction]:
         return self._table.transaction_versions()

--- a/python/tests/pyspark_integration/test_write_to_pyspark.py
+++ b/python/tests/pyspark_integration/test_write_to_pyspark.py
@@ -1,5 +1,6 @@
 """Tests that deltalake(delta-rs) can write to tables written by PySpark."""
 
+import os
 import pathlib
 
 import pyarrow as pa
@@ -162,6 +163,34 @@ def test_spark_read_z_ordered_history(tmp_path: pathlib.Path):
     dt.optimize.z_order(columns=["value"], partition_filters=[("id", "=", "1")])
 
     spark = get_spark()
+    history_df = spark.sql(f"DESCRIBE HISTORY '{tmp_path}'")
+
+    latest_operation_metrics = (
+        history_df.orderBy(history_df.version.desc()).select("operationMetrics").first()
+    )
+
+    assert latest_operation_metrics["operationMetrics"] is not None
+
+
+@pytest.mark.pyspark
+@pytest.mark.integration
+def test_spark_read_repair_run(tmp_path):
+    ids = ["1"] * 10
+    values = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
+
+    id_array = pa.array(ids, type=pa.string())
+    value_array = pa.array(values, type=pa.int32())
+
+    pa_table = pa.Table.from_arrays([id_array, value_array], names=["id", "value"])
+
+    write_deltalake(tmp_path, pa_table, mode="append")
+    write_deltalake(tmp_path, pa_table, mode="append")
+    dt = DeltaTable(tmp_path)
+    os.remove(dt.file_uris()[0])
+
+    dt.repair(dry_run=False)
+    spark = get_spark()
+
     history_df = spark.sql(f"DESCRIBE HISTORY '{tmp_path}'")
 
     latest_operation_metrics = (

